sycl/source/detail/graph_impl.cpp

```diff
@@ -296,6 +299,183 @@ void exec_graph_impl::makePartitions() {
+  std::vector<sycl::detail::CG *> FusionList;
+  std::vector<std::shared_ptr<node_impl>> NodesToBeFused;
+  bool CanAddNewKernels = true;
+  std::list<std::shared_ptr<node_impl>> HeadNonFusableNodes;
+  std::list<std::shared_ptr<node_impl>> RemainingNonFusableNodes;
+  for (auto &NodeImpl : DevicePartition->MSchedule) {
+    sycl::detail::CG *RawCG = NodeImpl->MCommandGroup.get();
+    // We check whether the graph is fusable when we create the FusionList.
+    // A graph is not fusable if a non-fusable CG is found in the middle of
+    // the serialized graph (i.e. MSchedule).
+    if (RawCG->getType() == sycl::detail::CGType::Kernel) {
+      if (!CanAddNewKernels) {
+        if (MRequireFusion) {
+          throw sycl::exception(
+              sycl::make_error_code(errc::feature_not_supported),
+              "The graph cannot be fused : Non-fusable nodes are in the middle "
+              "of the serialized graph.");
+        }
+        return;
+      }
+      FusionList.push_back(RawCG);
+      NodesToBeFused.push_back(NodeImpl);
+    } else if (FusionList.size() == 0) {
+      // Non fusable CG must be kept in the post fusion Schedule.
+      HeadNonFusableNodes.push_back(NodeImpl);
+
+    } else if (FusionList.size() > 0) {
+      // Non-fusable CG can be at the begining or the end of the graph.
+      // If a non fusable CG is found while CGs have been added to the
+      // FusionList, we set a flag to check that no other CG are added to
+      // FusionList after encountering this non-fusable CG.
+      CanAddNewKernels = false;
+      RemainingNonFusableNodes.push_back(NodeImpl);
+    }
+  }
+
+  // Call the JIT compiler to generate a new fused kernel.
+  auto FusedCG = sycl::detail::jit_compiler::get_instance().fuseKernels(
+      ContextImpl, DeviceImpl, FusionList, MFinalizePropList);
+
+  if (FusedCG == nullptr) {
+    if (MRequireFusion) {
+      throw sycl::exception(sycl::make_error_code(errc::feature_not_supported),
+                            "The graph cannot be fused.");
+    }
+    return;
+  }
+
+  // Re-create schedule with the new fused kernel
+  std::list<std::shared_ptr<node_impl>> FusedSchedule;
+  auto FusedNode =
+      std::make_shared<node_impl>(node_type::kernel, std::move(FusedCG));
+  // Add the dependencies of initial nodes to fused nodes
+  for (auto &Node : NodesToBeFused) {
+    constexpr auto CopyIfNotKernelAndNotPresent = [](const auto &NewNodes,
+                                                     auto &Nodes) {
+      std::copy_if(
+          NewNodes.begin(), NewNodes.end(), std::back_inserter(Nodes),
+          [&Nodes](auto &Node) {
+            return Node.lock()->MNodeType != node_type::kernel &&
+                   std::none_of(
+                       Nodes.begin(), Nodes.end(),
+                       [&Node](const std::weak_ptr<node_impl> &WeakNode) {
+                         return WeakNode.lock() == Node.lock();
+                       });
+          });
+    };
```

Comment:
preference for lifting this lambda outside of the for loop, and the one on line 427. Makes the indentation slightly better and wont shadow the `Node` variable